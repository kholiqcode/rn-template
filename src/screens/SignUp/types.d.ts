declare namespace SignUp {
  export type MainMenuScreenProp = StackNavigationProp<RootStackParamList, 'MainMenu'>;
  export interface IStyleSheet {
    textColor: string;
  }
}
