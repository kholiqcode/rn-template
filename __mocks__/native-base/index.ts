jest.mock('native-base', () => ({
  ...jest.requireActual('native-base'),
}));
