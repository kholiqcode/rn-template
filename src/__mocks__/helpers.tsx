// /* eslint-disable @typescript-eslint/no-explicit-any */
// // test-utils.jsx
// import React from 'react';
// import { render as rtlRender } from '@testing-library/react-native';
// import { configureStore } from '@reduxjs/toolkit';
// import { Provider } from 'react-redux';
// // Import your own reducer
// import userReducer from '../userSlice';

// function render(
//   ui: any,
//   {
//     preloadedState,
//     store),
//     ...renderOptions
//   }: { preloadedState: any; store: any; renderOptions: any } = {},
// ) {
//   function Wrapper({ children }: any) {
//     return <Provider store={store}>{children}</Provider>;
//   }
//   return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// }

// // re-export everything
// export * from '@testing-library/react-native';
// // override render method
// export { render };
