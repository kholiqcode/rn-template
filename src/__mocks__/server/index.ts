import { setupServer } from 'msw/node';
import { productsHandler, usersHandler } from './serverHandlers';

const server = setupServer(...productsHandler, ...usersHandler);

export { server };
