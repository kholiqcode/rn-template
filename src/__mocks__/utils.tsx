import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { store } from '../redux/index';

function renderWithProviders({ children }: PropsWithChildren<any>): JSX.Element {
  return <Provider store={store}>{children}</Provider>;
}

export { renderWithProviders };
