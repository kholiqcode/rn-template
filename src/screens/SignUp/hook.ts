import { useNavigation } from '@react-navigation/native';
import { useEffect, useState } from 'react';
import { Platform } from 'react-native';
import type { NativeStackNavigationProp } from '@react-navigation/native-stack';
// eslint-disable-next-line import/no-extraneous-dependencies
import { useForm } from 'react-hook-form';
import crashlytics from '@react-native-firebase/crashlytics';
import { call } from 'react-native-reanimated';

type NavigationProp = NativeStackNavigationProp<
  ReactNavigation.SignInStackParamList,
  'SignUpScreen'
>;
type FormData = {
  name: string;
  phone: string;
  email: string;
  password: string;
  passwordConfirmation: string;
};

export const useSignUp = () => {
  // STATE
  const [hidePassword, setHidePassword] = useState<boolean>(true);
  const form = useForm<FormData>();

  const navigation = useNavigation<NavigationProp>();
  const offsetKeyboard = Platform.select({
    ios: 0,
    android: 30,
  });

  // HANDLER
  const onToggleHidePassword = () => {
    setHidePassword(!hidePassword);
  };

  const onSubmit = async (data) => {
    try {
      crashlytics().log('User signed up.');
      crashlytics().crash();
      await Promise.all([
        crashlytics().setAttribute('phone', data.phone),
        crashlytics().setAttribute('email', data.email),
        crashlytics().setAttributes({
          password: data.password,
          passwordConfirmation: data.passwordConfirmation,
        }),
      ]);
      call.undefined.function();
    } catch (error: Error) {
      crashlytics().recordError(error);
    }
  };

  const goToSignIn = () => navigation.navigate('SignInScreen');

  // REACT HOOKS
  useEffect(() => {
    crashlytics().log('SignUp screen mounted.');

    return () => {
      crashlytics().log('SignUp screen unmounted.');
    };
  }, []);

  return { hidePassword, offsetKeyboard, onToggleHidePassword, goToSignIn, onSubmit, form };
};
