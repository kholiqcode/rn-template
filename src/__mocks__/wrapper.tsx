import React from 'react';
import { NativeBaseProvider } from 'native-base';
import { Provider } from 'react-redux';
import { store } from '../redux';

export const wrapper = ({ children }: { children: JSX.Element }) => {
  return (
    <NativeBaseProvider
      initialWindowMetrics={{
        frame: { x: 0, y: 0, width: 0, height: 0 },
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
      }}
    >
      <Provider store={store}>{children}</Provider>
    </NativeBaseProvider>
  );
};
