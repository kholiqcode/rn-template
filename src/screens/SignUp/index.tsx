import { Box, Button, Center, Input, KeyboardAvoidingView, Text } from 'native-base';
import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Controller } from 'react-hook-form';
import { useSignUp } from './hook';

export default function SignUp(): JSX.Element {
  const { hidePassword, offsetKeyboard, onSubmit, form } = useSignUp();

  return (
    <Box safeArea flex={1}>
      <KeyboardAvoidingView flex={1} keyboardVerticalOffset={offsetKeyboard}>
        <Center flex={1} px={10}>
          <Text fontSize={'3xl'} color={'blue.500'} mb={50}>
            SignUp
          </Text>
          <Controller
            control={form.control}
            render={({ field: { onChange, onBlur, value } }) => (
              <Input
                onBlur={onBlur}
                onChangeText={(val) => onChange(val)}
                value={value}
                placeholder="Type Name"
                mb="5"
                testID="input-name"
              />
            )}
            name="name"
            rules={{ required: 'Field is required', minLength: 3 }}
            defaultValue=""
          />
          <Controller
            control={form.control}
            render={({ field: { onChange, onBlur, value } }) => (
              <Input
                onBlur={onBlur}
                onChangeText={(val) => onChange(val)}
                value={value}
                placeholder="Type Phone"
                keyboardType="phone-pad"
                mb="5"
                testID="input-phone"
              />
            )}
            name="phone"
            rules={{ required: 'Field is required', minLength: 3 }}
            defaultValue=""
          />
          <Controller
            control={form.control}
            render={({ field: { onChange, onBlur, value } }) => (
              <Input
                onBlur={onBlur}
                onChangeText={(val) => onChange(val)}
                value={value}
                placeholder="Type Email"
                keyboardType="email-address"
                mb="5"
                testID="input-email"
              />
            )}
            name="email"
            rules={{ required: 'Field is required', minLength: 3 }}
            defaultValue=""
          />
          <Controller
            control={form.control}
            render={({ field: { onChange, onBlur, value } }) => (
              <Input
                onBlur={onBlur}
                onChangeText={(val) => onChange(val)}
                value={value}
                secureTextEntry={hidePassword}
                placeholder="Type Password"
                mb="5"
                testID="input-password"
              />
            )}
            name="password"
            rules={{ required: 'Field is required', minLength: 3 }}
            defaultValue=""
          />
          <Controller
            control={form.control}
            render={({ field: { onChange, onBlur, value } }) => (
              <Input
                onBlur={onBlur}
                onChangeText={(val) => onChange(val)}
                value={value}
                secureTextEntry={hidePassword}
                placeholder="Type Password Confimation"
                mb="5"
                testID="input-password-confirmation"
              />
            )}
            name="passwordConfirmation"
            rules={{ required: 'Field is required', minLength: 3 }}
            defaultValue=""
          />

          <Button
            onPress={form.handleSubmit(onSubmit)}
            variant={'solid'}
            size={'lg'}
            testID="btn-submit"
          >
            Submit
          </Button>
        </Center>
      </KeyboardAvoidingView>
    </Box>
  );
}
