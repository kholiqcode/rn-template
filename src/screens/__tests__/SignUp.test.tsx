import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { wrapper } from '../../../__mocks__/wrapper';
import SignUp from '../SignUp';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: mockedNavigate }),
}));

describe('SignUp screen', () => {
  const data = {
    name: 'Kholiq',
    phone: '0818181818',
    email: 'abdul@finetiks.com',
    password: 'test123',
    passwordConfirmation: 'test123',
  };
  it('should can input form', async () => {
    const rendered = render(<SignUp />, { wrapper });

    const inputName = rendered.getByTestId('input-name');
    fireEvent.changeText(inputName, data.name);
    const inputPhone = rendered.getByTestId('input-phone');
    fireEvent.changeText(inputPhone, data.phone);
    const inputEmail = rendered.getByTestId('input-email');
    fireEvent.changeText(inputEmail, data.email);
    const inputPassword = rendered.getByTestId('input-password');
    fireEvent.changeText(inputPassword, data.password);
    const inputPasswordConfirmation = rendered.getByTestId('input-password-confirmation');
    fireEvent.changeText(inputPasswordConfirmation, data.passwordConfirmation);

    const btnSubmit = rendered.getByTestId('btn-submit');
    fireEvent.press(btnSubmit);

    expect(rendered).toMatchSnapshot();
  });
});
