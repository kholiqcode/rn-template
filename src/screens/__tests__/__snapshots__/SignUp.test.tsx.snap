// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp screen should can input form 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      dataSet={Object {}}
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        dataSet={Object {}}
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "paddingLeft": 40,
            "paddingRight": 40,
          }
        }
      >
        <Text
          dataSet={Object {}}
          style={
            Object {
              "backgroundColor": undefined,
              "color": "#3b82f6",
              "fontFamily": undefined,
              "fontSize": 30,
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 45,
              "marginBottom": 50,
              "textDecorationLine": undefined,
            }
          }
        >
          SignUp
        </Text>
        <View
          dataSet={Object {}}
          isFocused={false}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#d4d4d4",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "marginBottom": 20,
              "overflow": "hidden",
            }
          }
        >
          <TextInput
            accessible={true}
            dataSet={Object {}}
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="Type Name"
            placeholderTextColor="#a3a3a3"
            secureTextEntry={false}
            selectionColor="#1f2937"
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#171717",
                "flex": 1,
                "fontFamily": undefined,
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "height": "100%",
                "paddingBottom": 8,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 8,
                "width": "100%",
              }
            }
            testID="input-name"
            value="Kholiq"
          />
        </View>
        <View
          dataSet={Object {}}
          isFocused={false}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#d4d4d4",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "marginBottom": 20,
              "overflow": "hidden",
            }
          }
        >
          <TextInput
            accessible={true}
            dataSet={Object {}}
            editable={true}
            keyboardType="phone-pad"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="Type Phone"
            placeholderTextColor="#a3a3a3"
            secureTextEntry={false}
            selectionColor="#1f2937"
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#171717",
                "flex": 1,
                "fontFamily": undefined,
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "height": "100%",
                "paddingBottom": 8,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 8,
                "width": "100%",
              }
            }
            testID="input-phone"
            value="0818181818"
          />
        </View>
        <View
          dataSet={Object {}}
          isFocused={false}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#d4d4d4",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "marginBottom": 20,
              "overflow": "hidden",
            }
          }
        >
          <TextInput
            accessible={true}
            dataSet={Object {}}
            editable={true}
            keyboardType="email-address"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="Type Email"
            placeholderTextColor="#a3a3a3"
            secureTextEntry={false}
            selectionColor="#1f2937"
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#171717",
                "flex": 1,
                "fontFamily": undefined,
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "height": "100%",
                "paddingBottom": 8,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 8,
                "width": "100%",
              }
            }
            testID="input-email"
            value="abdul@finetiks.com"
          />
        </View>
        <View
          dataSet={Object {}}
          isFocused={false}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#d4d4d4",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "marginBottom": 20,
              "overflow": "hidden",
            }
          }
        >
          <TextInput
            accessible={true}
            dataSet={Object {}}
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="Type Password"
            placeholderTextColor="#a3a3a3"
            secureTextEntry={true}
            selectionColor="#1f2937"
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#171717",
                "flex": 1,
                "fontFamily": undefined,
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "height": "100%",
                "paddingBottom": 8,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 8,
                "width": "100%",
              }
            }
            testID="input-password"
            value="test123"
          />
        </View>
        <View
          dataSet={Object {}}
          isFocused={false}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#d4d4d4",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "marginBottom": 20,
              "overflow": "hidden",
            }
          }
        >
          <TextInput
            accessible={true}
            dataSet={Object {}}
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="Type Password Confimation"
            placeholderTextColor="#a3a3a3"
            secureTextEntry={true}
            selectionColor="#1f2937"
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#171717",
                "flex": 1,
                "fontFamily": undefined,
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "400",
                "height": "100%",
                "paddingBottom": 8,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 8,
                "width": "100%",
              }
            }
            testID="input-password-confirmation"
            value="test123"
          />
        </View>
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          colorScheme="primary"
          dataSet={Object {}}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#0891b2",
              "borderRadius": 4,
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
            }
          }
          testID="btn-submit"
        >
          <View
            dataSet={Object {}}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
            test={true}
          >
            <View
              dataSet={Object {}}
              style={Object {}}
            >
              <Text
                dataSet={Object {}}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "color": "#fafafa",
                    "fontFamily": undefined,
                    "fontSize": 16,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 24,
                    "textDecorationLine": undefined,
                  }
                }
              >
                Submit
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
